File: FrontendRazorPage/Core/Services/GrammarClientService.cs
using FrontendRazorPage.Core.Services;
using FrontendRazorPage.Models;
using FrontendRazorPage.Models.AdminModel;
using System.Net.Http;
using System.Net.Http.Json;
using static System.Net.WebRequestMethods;

namespace FrontendRazorPage.Services
{
    public class GrammarClientService : BaseClientService
    {

       
        private readonly string _apiBase = "api/grammar";

        public GrammarClientService(HttpClient httpClient) : base(httpClient) { }
        

        public async Task<GrammarModel> GetGrammarByIdAsync(int grammarId)
        {
            if (grammarId <= 0) return new GrammarModel();
            try
            {
                // Bắn lệnh GET sang endpoint: /api/Grammar/{grammarId}
                // Nếu lỗi mạng hoặc null, toán tử ?? new() sẽ trả về object trống để chống sập giao diện
                return await _httpClient.GetFromJsonAsync<GrammarModel>($"{_apiBase}/{grammarId}") ?? new();
            }
            catch
            {
                // Nếu Backend sập hoặc mất mạng, trả về object trống để app chạy tiếp mượt mà
                return new GrammarModel();
            }
        }

        public async Task<List<GrammarModel>> GetGrammarByLessonAsync(int lessonId)
        {
            if (lessonId <= 0) return new List<GrammarModel>();

            try
            {
                // Bắn lệnh GET sang endpoint: /api/Grammar/lesson/{lessonId}
                // Nếu lỗi mạng hoặc null, toán tử ?? new() sẽ trả về danh sách trống để chống sập giao diện
                return await _httpClient.GetFromJsonAsync<List<GrammarModel>>($"{_apiBase}/lesson/{lessonId}") ?? new();
            }
            catch
            {
                // Nếu Backend sập hoặc mất mạng, trả về list trống để app chạy tiếp mượt mà
                return new List<GrammarModel>();
            }
        }

        public async Task<List<QuestionModel>> GetQuestionsByGrammarAsync(int grammarId, int questionType = 0)
        {
            if (grammarId <= 0) return new List<QuestionModel>();

            try
            {
                // Bắn URL dạng số: /api/Grammar/questions?grammarId=1&questionType=1
                string url = $"{_apiBase}/questions?grammarId={grammarId}&questionType={questionType}";
                return await _httpClient.GetFromJsonAsync<List<QuestionModel>>(url) ?? new();
            }
            catch
            {
                return new List<QuestionModel>();
            }
        }



        /// ////////////////////////////ADMIN//////////////////////////////



        // 1. Lấy danh sách cho Admin 
        public async Task<List<GrammarAdminModel>> GetAllForAdminAsync()
        {
            try
            {
                // Sửa thành đường dẫn tuyệt đối chuẩn xác kết hợp với _apiBase
                return await _httpClient.GetFromJsonAsync<List<GrammarAdminModel>>($"{_apiBase}/admin-list") ?? new();
            }
            catch
            {
                return new List<GrammarAdminModel>();
            }
        }

        // 2. Lấy chi tiết theo ID cho Admin
        public async Task<GrammarAdminModel?> GetByIdForAdminAsync(int id)
        {
            try
            {
                return await _httpClient.GetFromJsonAsync<GrammarAdminModel>($"{_apiBase}/admin/{id}");
            }
            catch
            {
                return null;
            }
        }

        public async Task<bool> CreateAsync(GrammarAdminModel grammar)
        {
            try
            {
                var response = await _httpClient.PostAsJsonAsync($"{_apiBase}", grammar);
                return response.IsSuccessStatusCode;
            }
            catch
            {
                return false;
            }
        }


        // 4. Cập nhật mẫu ngữ pháp
        public async Task<bool> UpdateAsync(int id, GrammarAdminModel grammar)
        {
            try
            {
                var response = await _httpClient.PutAsJsonAsync($"{_apiBase}/{id}", grammar);
                return response.IsSuccessStatusCode;
            }
            catch
            {
                return false;
            }
        }

        // 5. Xóa mẫu ngữ pháp
        public async Task<bool> DeleteAsync(int id)
        {
            try
            {
                var response = await _httpClient.DeleteAsync($"{_apiBase}/{id}");
                return response.IsSuccessStatusCode;
            }
            catch
            {
                return false;
            }
        }
    }

}

File: FrontendAngular/src/app/Core/Services/grammar-client-service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, of } from 'rxjs';
import { catchError, map } from 'rxjs/operators';
//service
import { BaseService } from './BaseService';
//Model
import { GrammarModel } from '../../Models/grammar-model';
import { GrammarAdminModel } from '../../Models/AdminModel/grammar-admin-model'; // Đảm bảo bạn đã tạo model này


@Injectable({
    providedIn: 'root'
})
export class GrammarClientService extends BaseService {
    // api/grammar
    private readonly _apiBase = `${this._apiBaseUrl}/grammar`;

    constructor(private _http: HttpClient) {
        super(); // Bắt buộc khi kế thừa BaseService
    }

    /**
     * Lấy chi tiết ngữ pháp theo ID cho User
     * Tương ứng C#: GetGrammarByIdAsync
     */
    public getGrammarByIdAsync(grammarId: number): Observable<GrammarModel> {
        if (grammarId <= 0) return of({} as GrammarModel);

        return this._http.get<GrammarModel>(`${this._apiBase}/${grammarId}`).pipe(
            catchError(() => of({} as GrammarModel)) // Trả về object trống nếu lỗi mạng/backend sập
        );
    }

    /**
     * Lấy danh sách ngữ pháp theo bài học (Lesson)
     * Tương ứng C#: GetGrammarByLessonAsync
     */
    public getGrammarByLessonAsync(lessonId: number): Observable<GrammarModel[]> {
        if (lessonId <= 0) return of([]);

        return this._http.get<GrammarModel[]>(`${this._apiBase}/lesson/${lessonId}`).pipe(
            catchError(() => of([])) // Trả về list trống nếu lỗi mạng
        );
    }

    /// //////////////////////////// ADMIN //////////////////////////////

    /**
     * 1. Lấy danh sách tất cả ngữ pháp cho Admin
     * Tương ứng C#: GetAllForAdminAsync
     */
    public getAllForAdminAsync(): Observable<GrammarAdminModel[]> {
        return this._http.get<GrammarAdminModel[]>(`${this._apiBase}/admin-list`).pipe(
            catchError(() => of([]))
        );
    }

    /**
     * 2. Lấy chi tiết ngữ pháp theo ID cho Admin
     * Tương ứng C#: GetByIdForAdminAsync
     */
    public getByIdForAdminAsync(id: number): Observable<GrammarAdminModel | null> {
        return this._http.get<GrammarAdminModel>(`${this._apiBase}/admin/${id}`).pipe(
            catchError(() => of(null)) // Trả về null nếu xảy ra lỗi giống C#
        );
    }


    /**
     * 3. Thêm mới mẫu ngữ pháp
     * Tương ứng C#: CreateAsync
     */
    public createAsync(grammar: GrammarAdminModel): Observable<boolean> {
        return this._http.post(`${this._apiBase}`, grammar).pipe(
            map(() => true),             // Nếu thành công (2xx), ép kết quả về true
            catchError(() => of(false))  // Nếu lỗi mạng/backend sập, trả về false
        );
    }

    /**
     * 4. Cập nhật mẫu ngữ pháp
     * Tương ứng C#: UpdateAsync
     */
    public updateAsync(id: number, grammar: GrammarAdminModel): Observable<boolean> {
        return this._http.put(`${this._apiBase}/${id}`, grammar).pipe(
            map(() => true),
            catchError(() => of(false))
        );
    }

    /**
     * 5. Xóa mẫu ngữ pháp
     * Tương ứng C#: DeleteAsync
     */
    public deleteAsync(id: number): Observable<boolean> {
        return this._http.delete(`${this._apiBase}/${id}`).pipe(
            map(() => true),
            catchError(() => of(false))
        );
    }
}